nixos/redshift: use config file instead of passing parameters

thiagokokada opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Not every option is exposed by `redshift`/`gammastep` parameters, for example gamma options are only exposed in configuration file. So this PR refactors this module to generate a configuration file and pass it to the `redshift`/`gammastep` using -c parameter.

Also, it implements fade/gamma options not available in the previous version of this module (since it was not exposed by parameters).

This is a slightly breaking change, since `extraOptions` was renamed to `extraConfig` and now accept a attribute set instead of a list of string

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

thiagokokada committed:
nixos/redshift: use config file instead of passing parameters

Not every option is exposed by redshift/gammastep parameters, for
example gamma options are only exposed in configuration file. So this PR
refactors this module to generate a configuration file and pass it to
the redshift/gammastep using -c parameter.

Also, it implements fade/gamma options not available in the previous
version of this module (since it was not exposed by parameters).

This is a slightly breaking change, since extraOptions was renamed to
extraConfig and now accept a attribute set instead of a list of strings.

thiagokokada commented:
With the following configuration:

```nix
let
    redshiftModuleRefactor = fetchTarball {
    url = "https://github.com/thiagokokada/nixpkgs/archive/1bea263ca0937e4d7dd08b85545a2426223c7a79.tar.gz";
    sha256 = "0gawqndyj2n23mhsp93rlacrfb8x86rr6hd7d9waknh6fpqbnixr";
  };
in{
  # Backport module from unstable.
  imports = [
    "${redshiftModuleRefactor}/nixos/modules/services/x11/redshift.nix"
  ];

  disabledModules = [
    "services/x11/redshift.nix"
  ];

  location = {
    latitude = -23.5;
    longitude = -46.6;
  };

  services = {
    redshift = {
      enable = true;
      temperature = {
        day = 5500;
        night = 3700;
      };
      gamma = {
        day = 1.0;
        night = 0.8;
      };
      package = pkgs.unstable.gammastep;
      executable = "/bin/gammastep-indicator";
    };
  };
}
```

I had the following result:

```sh
~
❯ systemctl cat --user redshift.service     
# /nix/store/8bxr1gi8y9qa8ya4qya88vx97mzdm6g3-unit-redshift.service/redshift.service
[Unit]
Description=Redshift colour temperature adjuster
PartOf=graphical-session.target

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/a2px4kdz1jm03f8ifr1pzir0csfmyrlv-glibc-locales-2.31/lib/locale/locale-archive"
Environment="PATH=/nix/store/z1qvlavy35wanw5k54fvvfffws5bvigj-coreutils-8.31/bin:/nix/store/3fvzxz59gacagpwyzpfdiinc1yv46hw1-findutils-4>
Environment="TZDIR=/nix/store/w1g27pgslf28nh1py1szj7lk4xksdhqq-tzdata-2020c/share/zoneinfo"



ExecStart=/nix/store/f38nk3a9qq415zhar429j7fr7czarvcg-gammastep-2.0.6/bin/gammastep-indicator -c /nix/store/xswxcml4r3fdzviyqwd399lpzsia>
Restart=always
RestartSec=3


~ 12s
❯ cat /nix/store/xswxcml4r3fdzviyqwd399lpzsia3v27-redshift.conf 
[general]
temp-day=5500
temp-night=3700
fade=1
location-provider=manual
brightness-day=1.000000
brightness-night=1.000000
gamma-day=1.000000
gamma-night=0.800000



[manual]
lat=-23.500000
lon=-46.600000
```

For `redshift` case:

```nix
let
    redshiftModuleRefactor = fetchTarball {
    url = "https://github.com/thiagokokada/nixpkgs/archive/1bea263ca0937e4d7dd08b85545a2426223c7a79.tar.gz";
    sha256 = "0gawqndyj2n23mhsp93rlacrfb8x86rr6hd7d9waknh6fpqbnixr";
  };
in{
  # Backport module from unstable.
  imports = [
    "${redshiftModuleRefactor}/nixos/modules/services/x11/redshift.nix"
  ];

  disabledModules = [
    "services/x11/redshift.nix"
  ];

  location.provider = "geoclue2";

  services = {
    redshift = {
      enable = true;
      temperature = {
        day = 5500;
        night = 3700;
      };
      gamma = {
        day = 1.0;
        night = 0.8;
      };
      package = pkgs.unstable.redshift;
      executable = "/bin/redshift-gtk";
    };
  };
}
```

```sh
❯ systemctl cat --user redshift.service
# /nix/store/9xqyys1w2lb6f9wma2zjjfyhr4hsjsc7-unit-redshift.service/redshift.service
[Unit]
Description=Redshift colour temperature adjuster
PartOf=graphical-session.target

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/a2px4kdz1jm03f8ifr1pzir0csfmyrlv-glibc-locales-2.31/lib/locale/locale-archive"
Environment="PATH=/nix/store/z1qvlavy35wanw5k54fvvfffws5bvigj-coreutils-8.31/bin:/nix/store/3fvzxz59gacagpwyzpfdiinc1yv46hw1-findutils-4>
Environment="TZDIR=/nix/store/w1g27pgslf28nh1py1szj7lk4xksdhqq-tzdata-2020c/share/zoneinfo"



ExecStart=/nix/store/kx8r729wmn9nsm0plqiqv6gvkiah8jwy-redshift-1.12/bin/redshift-gtk -c /nix/store/6ffjf4pbb68ffx6xr0hy2dcjg5rkyi3x-reds>
Restart=always
RestartSec=3


~
❯ cat /nix/store/6ffjf4pbb68ffx6xr0hy2dcjg5rkyi3x-redshift.conf 
[redshift]
temp-day=5500
temp-night=3700
fade=1
location-provider=geoclue2
brightness-day=1.000000
brightness-night=1.000000
gamma-day=1.000000
gamma-night=0.800000





```

thiagokokada commented:
CC @yegortimoshenko @globin @primeos (redshift/gammastep package maintainers).

Also @sumnerevans @Infinisil @michaelpj (last people to touch this module).

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -92,12 +119,20 @@ in {
-      example = [ "-v" "-m randr" ];
+      type = types.str;
+      default = "";
+      example = ''
+        ; Will be applied in "[redshift]" section
+        elevation-high=3
+        elevation-low=-6
+
+        ; Using another section
+        [randr]
+        screen=0
+      '';
```
I am not sure if this is the best way to do it. I think that maybe accepting a map, something like:

```nix
{
  randr = {
    screen = "0";
  };
}
```

Maybe would be better but probably way more complicated (I don't have a clue on how to implement this in Nix). Also I don't know what to do in the conflict cases (`{ redshift = { ... }` since there is already a `[redshift]` section).

thiagokokada replied:
BTW, if this way I did is ok I can document this breaking change in release notes.

sumnerevans replied:
Look at the `toINI` generator. It writes dictionaries to various formats INI. I think this simplifies everything, because then you can leave everything as Nix sets until the very end where you write to the file.

thiagokokada replied:
@sumnerevans Does it solve the conflict cases too :thinking: ? Like:

```nix
{
  foo =  {
    bar = "1";
    qux = "1";
  };
  foo = {
    baz = "2";
    qux = "3";
  };
}
```

```
[foo]
bar = "1"
baz = "2"
qux = "3"
```

?

sumnerevans replied:
I think that the Nix `//` operator should do that... Like:
```
{
  foo =  {
    bar = "1";
    qux = "1";
  };
} // {
  foo = {
    baz = "2";
    qux = "3";
  };
}
```

then you can just write that using the `toINI` function. Not at my Nix computer (on my work computer right now), but I think that something like that should work.

thiagokokada replied:
Answering my question:

```
nix-repl> lib.generators.toINI {} { foo = { bar = "qux"; }; foo = { baz = "quux"; }; }  
"[foo]\nbar=qux\nbaz=quux\n"
nix-repl> lib.generators.toINI {} { foo = { bar = "qux"; }; foo = { bar = "quux"; baz = "quux"; }; }    
error: attribute 'bar' at (string):1:35 already defined at (string):1:59
```

So defining the same root key is fine, redefining a key is not. I think this is ok I guess, will work sufficiently fine for my purposes.

sumnerevans replied:
OK, that looks great! BTW, thanks for doing this, definitely better to have it as a config file IMO.

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -113,22 +148,42 @@ in {
+        gamma-day=${toString cfg.gamma.day}
+        gamma-night=${toString cfg.gamma.night}
+
+        ${cfg.extraOptions}
+
+        ${locationString}
```
Probably there is a better way to do it where it wouldn't leave blank spaces if the option is blank, but anyway this doesn't affect the result so I didn't really bother.

sumnerevans replied:
This is solved if you use the `toINI` generator.

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -56,23 +56,50 @@ in {
       day = mkOption {
-        type = types.str;
-        default = "1";
+        type = types.float;
+        default = 1.0;
```
This is also technically a breaking change. I can document it in release notes if necessary.

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -113,22 +148,42 @@ in {
+        ''
+        else "";
+      fadeString = if cfg.fade then "1" else "0";
+      sectionString = if strings.hasInfix "gammastep" cfg.executable
+        # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+        then "[general]"
+        else "[redshift]";
```
Not really necessary since `gammastep` still supports `[redshift]`, but this avoids a warning at startup.

thiagokokada commented:
With the changes in 460e276, the following config file:

```nix
{
  location.provider = "geoclue2";

  services = {
    redshift = {
      enable = true;
      temperature = {
        day = 5500;
        night = 3700;
      };
      gamma = {
        day = 1.0;
        night = 0.8;
      };
      package = pkgs.unstable.redshift;
      executable = "/bin/redshift-gtk";
      extraConfig = {
        randr = {
          display = 0;
        };
     };
    };
  };
}
```
```
[randr]
screen=0

[redshift]
brightness-day=1.000000
brightness-night=1.000000
fade=1
gamma-day=1.000000
gamma-night=0.800000
location-provider=geoclue2
temp-day=5500
temp-night=3700
```

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -91,13 +121,27 @@ in {
-      type = types.listOf types.str;
-      default = [];
-      example = [ "-v" "-m randr" ];
+    extraConfig = mkOption {
```
Renamed because this is technically not a option anymore.

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -106,31 +150,51 @@ in {
+        ExecStart = "${cfg.package}${cfg.executable} -c ${configFile}";
         RestartSec = 3;
-        Restart = "always";
+        Restart = "on-failure";
```
This avoids the service to keep restarting in case of an error in config (that may happen for example when setting a wrong `extraConfig` value).

eadwu commented:
I don't believe I'm a maintainer for redshift nor do I use redshift locally anymore.

Anyway if you're going to generate a configuration file it might be better to convert it to use a structured attribute set (i.e. `settings`) and generate the file with contents converted from `generators.toINI`.

thiagokokada commented:
@eadwu This is exactly what I did in commit 460e276.

eadwu reviewed (approved):
Looks fine to me. Not quite sure what about the gamma configuration though, it looks like it needs an external dependency (`gammastep`) to work, though if it doesn't break the configuration parsing it's probably fine (though it will probably fall under unexpected behavior).

Nothing else for me to note except nitpicks which probably don't matter. Doesn't break system evaluation on my end.

You can probably put the entire configuration inside `extraConfig` but I don't think it would provide any extra benefit unless you are using `nixos-option` to view the attribute set.

thiagokokada commented:
> Not quite sure what about the gamma configuration though, it looks like it needs an external dependency (gammastep) to work

`gamma-day` and `gamma-night` are available in `redshift`: https://github.com/jonls/redshift/blob/master/redshift.conf.sample#L39-L40.

BTW, `gammastep` (https://gitlab.com/chinstrap/gammastep/) is a fork of `redshift` that supports Wayland, not a dependency of `redshift`. It supports pretty much the same settings. It is the recommended fork for `sway` (AFAIK) after `redshift-wlr` was abandoned.

xaverdh commented:
see https://github.com/NixOS/nixpkgs/pull/50979 for prior art

thiagokokada committed:
nixos/redshift: allow setting gamma.{day,night} as a string

Redshift support setting each color channel individually in the
following format: "0.8:07:0.8". So support this case too.

Also, improve type checking in `extraConfig`.

thiagokokada commented:
Commit 36156ba adds support for setting `gamma-{day,night} = "0.8:0.7:0.8"` to support this use case too. Also improves type checking in `extraConfig`.

I think this PR is pretty much ready to merge.

> see #50979 for prior art

Huh... I think this other proposal is kinda abandoned? Well, if the original author still wants to merge I am okay with it too, but I think this PR is ready to merge and the other one still needs some fixes right?

aanderse commented:
A few conventions that we use would include calling `extraOptions` `settings`, using `(pkgs.formats.ini {}).format` for `settings.type`, using `(pkgs.formats.ini {}).generate` to generate to configuration file, and using `mkMerge` when appropriate. As mentioned, please see RFC42 for details.

thiagokokada commented:
@aanderse Didn't know we had a RFC for this (it is not merged and I didn't find it in https://github.com/NixOS/rfcs/, had to look on the open PRs). TIL.

I was trying to keep this module was similar as the current module breaking as little as possible, but I think this RFC points is fair. Following it would mean probably a completely different module as the current one, and I think this is kinda what #50979 is.

Well, will follow the #50979 PR for now and if the author doesn't answer in one or two weeks, I can probably try to open another PR following more closely RFC 42.

thiagokokada commented:
BTW, should I close this PR or is there still interest in merging it?

aanderse commented:
> BTW, should I close this PR or is there still interest in merging it?

Let's wait and see if the other PR picks up or not.

thiagokokada commented:
> Let's wait and see if the other PR picks up or not.

Asking more because I won't adapt this PR to follow RFC 42 since this will probably would be better made in another PR (it will be a full rewrite anyway). So if RFC 42 support is a must I don't see many reasons to keep this open.

aanderse commented:
Please **never** mistake my opinions for a "must" :laughing:

thiagokokada commented:
Ok, so I will keep this open and if the maintainers of the module thinks this is good enough they can go and approve/merge.

BTW, for me it is in a good enough state (RFC 42 really seems to make things more elegant, but I am using this module in my configuration and it works good enough for me).

thiagokokada commented:
Substitute PR: https://github.com/NixOS/nixpkgs/pull/108739.
